File: clients/java/src/main/java/MyStrategy.java
import model.*;

public class MyStrategy {
  static double distanceSqr(Vec2Double a, Vec2Double b) {
    return (a.getX() - b.getX()) * (a.getX() - b.getX()) + (a.getY() - b.getY()) * (a.getY() - b.getY());
  }

  public UnitAction getAction(Unit unit, Game game, Debug debug) {
    Unit nearestEnemy = null;
    for (Unit other : game.getUnits()) {
      if (other.getPlayerId() != unit.getPlayerId()) {
        if (nearestEnemy == null || distanceSqr(unit.getPosition(),
            other.getPosition()) < distanceSqr(unit.getPosition(), nearestEnemy.getPosition())) {
          nearestEnemy = other;
        }
      }
    }
    LootBox nearestWeapon = null;
    for (LootBox lootBox : game.getLootBoxes()) {
      if (lootBox.getItem() instanceof Item.Weapon) {
        if (nearestWeapon == null || distanceSqr(unit.getPosition(),
            lootBox.getPosition()) < distanceSqr(unit.getPosition(), nearestWeapon.getPosition())) {
          nearestWeapon = lootBox;
        }
      }
    }
    Vec2Double targetPos = unit.getPosition();
    if (unit.getWeapon() == null && nearestWeapon != null) {
      targetPos = nearestWeapon.getPosition();
    } else if (nearestEnemy != null) {
      targetPos = nearestEnemy.getPosition();
    }
    debug.draw(new CustomData.Log("Target pos: " + targetPos));
    Vec2Double aim = new Vec2Double(0, 0);
    if (nearestEnemy != null) {
      aim = new Vec2Double(nearestEnemy.getPosition().getX() - unit.getPosition().getX(),
          nearestEnemy.getPosition().getY() - unit.getPosition().getY());
    }
    boolean jump = targetPos.getY() > unit.getPosition().getY();
    if (targetPos.getX() > unit.getPosition().getX() && game.getLevel()
        .getTiles()[(int) (unit.getPosition().getX() + 1)][(int) (unit.getPosition().getY())] == Tile.WALL) {
      jump = true;
    }
    if (targetPos.getX() < unit.getPosition().getX() && game.getLevel()
        .getTiles()[(int) (unit.getPosition().getX() - 1)][(int) (unit.getPosition().getY())] == Tile.WALL) {
      jump = true;
    }
    UnitAction action = new UnitAction();
    action.setVelocity(targetPos.getX() - unit.getPosition().getX());
    action.setJump(jump);
    action.setJumpDown(!jump);
    action.setAim(aim);
    action.setShoot(true);
    action.setSwapWeapon(false);
    action.setPlantMine(false);
    return action;
  }
}

File: clients/javascript/my-strategy.js
const UnitAction = require('./model/unit-action').UnitAction;
const Vec2Double = require('./model/vec2-double').Vec2Double;
const Item = require('./model/item').Item;
const Tile = require('./model/tile');
const CustomData = require('./model/custom-data').CustomData;

class MyStrategy {
    async getAction (unit, game, debug) {
        const distanceSqr = function (a, b) {
            return Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2);
        };

        let minDistance = Number.POSITIVE_INFINITY;
        const nearestEnemy = game.units
            .filter((u) => {
                return u.playerId !== unit.playerId;
            })
            .reduce(function (prev, u) {
                let currentDistance = distanceSqr(u.position, unit.position);
                if (currentDistance < minDistance) {
                    minDistance = currentDistance;
                    return u;
                }
                return prev;
            });
        
        minDistance = Number.POSITIVE_INFINITY;
        const nearestWeapon = game.lootBoxes
            .filter ((box) => {
                return box.item instanceof Item.Weapon;
            })
            .reduce(function (prev, box) {
                let currentDistance = distanceSqr(box.position, unit.position);
                if (currentDistance < minDistance) {
                    minDistance = currentDistance;
                    return box;
                }
                return prev;
            });

        let targetPos = unit.position;
        if (unit.weapon === null && nearestWeapon) {
            targetPos = nearestWeapon.position;
        } else {
            targetPos = nearestEnemy.position;
        }
        await debug.draw(new CustomData.Log(`Target pos: ${targetPos.toString()}`));
        
        let aim = new Vec2Double(0, 0);
        if (nearestEnemy) {
            aim = new Vec2Double(
                nearestEnemy.position.x - unit.position.x,
                nearestEnemy.position.y - unit.position.y
            );
        }

        let jump = targetPos.y > unit.position.y;
        if (targetPos.x > unit.position.x && game.level.tiles[parseInt(unit.position.x + 1)][parseInt(unit.position.y)] === Tile.Wall) {
            jump = true;
        }
            
        if (targetPos.x < unit.position.x && game.level.tiles[parseInt(unit.position.x - 1)][parseInt(unit.position.y)] === Tile.Wall) {
            jump = true;
        }

        return new UnitAction(
            targetPos.x - unit.position.x,
            jump,
            !jump,
            aim,
            true,
            false,
            false,
            false
        );
    }
}

module.exports.MyStrategy = MyStrategy;
